craft_parts/state_manager/state_manager.py

```diff
@@ -0,0 +1,152 @@
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Part crafter step state management."""
+
+import itertools
+from dataclasses import dataclass
+from typing import Dict, Optional
+
+from craft_parts.steps import Step
+
+from .states import StepState
+
+
+@dataclass(frozen=True)
+class _StateWrapper:
+    """A wrapper for the in-memory StepState class with extra metadata.
+
+    This is a wrapper class for StepState that adds additional metadata
+    such as the update sequence order to the data loaded from a previous
+    lifecycle run. Metadata is read-only to prevent unintentional changes.
+    """
+
+    state: StepState
+    serial: int
+    step_updated: bool = False
+
+    def is_newer_than(self, other: "_StateWrapper"):
+        """Verify if this state is newer than the specified state.
+
+        :param other: The wrapped state to compare this state to.
+        """
+        return self.serial > other.serial
+
+
+class _EphemeralStates:
+    """A dictionary-backed simple database manager for wrapped states."""
+
+    def __init__(self):
+        self._state: Dict[str, Dict[Step, _StateWrapper]] = {}
+        self._serial_gen = itertools.count(1)
+
+    def wrap_state(self, state: StepState, step_updated: bool = False) -> _StateWrapper:
+        """Add metadata to step state.
+
+        :param state: The part state to store.
+        :param step_updated: Whether this state was updated after an
+            outdated report.
+
+        :return: The wrapped state with additional metadata.
+        """
+        stw = _StateWrapper(
+            state, serial=next(self._serial_gen), step_updated=step_updated
+        )
+        return stw
+
+    def set(
+        self, *, part_name: str, step: Step, state: Optional[_StateWrapper]
+    ) -> None:
+        """Set a state for a given part and step.
+
+        :param part_name: The name of the part corresponding to the state
+            to be set.
+        :param step: The step corresponding to the state to be set.
+        :param state: The state to assign to the part name and step.
+        """
+        if not state:
+            self.remove(part_name=part_name, step=step)
+            return
+
+        if part_name not in self._state:
+            self._state[part_name] = dict()
+
+        self._state[part_name][step] = state
+
+    def get(self, *, part_name: str, step: Step) -> Optional[_StateWrapper]:
+        """Retrieve the state for a given part and step.
+
+        :param part_name: The name of the part corresponding to the state
+            to be retrieved.
+        :param step: The step corresponding to the state to be retrieved.
+
+        :return: The wrapped state assigned to the part name and step.
+        """
+        if self.test(part_name=part_name, step=step):
+            return self._state[part_name][step]
+        return None
+
+    def test(self, *, part_name: str, step: Step) -> bool:
```

Comment:
```suggestion
    def has_part_step_state(self, *, part_name: str, step: Step) -> bool:
```

Or perhaps replace with
```suggestion
    def get_part_step_state(self, *, part_name: str, step: Step) -> Optional[_SttateWrapper]:
```